### PR TITLE
[sublimetext4] Add subl.exe shim, improve package

### PR DESCRIPTION
**Pull request by koitsu**
Hopefully will address [Jackenmen/choco-auto/issues/6](https://github.com/Jackenmen/choco-auto/issues/6)

- Add: call ShimGen for subl.exe, so you can now use it via CLI.

- Bug: Properly detect Sublime Text uninstaller registry keys.  Sublime Text 4 registry key has been named 'Sublime Text_is1' ever since they switch to Inno Setup installer.  Older Sublime Text installs (3.x and pre-InnoSetup-4.x) might have other registry keys though.  As such, key off of 'Sublime Text*' (to match all possibilities).  Note: this fix is needed for the subl.exe ShimGen to work correctly!

- Bug: No need for 90% of chocolateyUninstall.ps1 given the introduction of Chocolatey's Auto Uninstaller (enabled since Chocolatey 0.9.10). If a user has disabled feature, then they probably don't want the uninstaller running anyway.

### CODE REVIEW

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 [hunk 1]
```diff
@@ -1,26 +1,3 @@
 $ErrorActionPreference = 'Stop'
 
-$packageName = "sublimetext4"
-
-[array]$key = Get-UninstallRegistryKey -SoftwareName "Sublime Text 4"
```
Comment (Jackenmen) [inline automatic/sublimetext4/tools/chocolateyUninstall.ps1:5]: The code for uninstallation is required because `QuietUninstallString` contains `"C:\Program Files\Sublime Text\unins000.exe" /SILENT` which isn't enough for the uninstall to be fully silent - the uninstall window still pops up.

However, as you've noticed, Sublime Text 4's installer has been changed to Inno Setup which seems to have made this code do nothing due to the software name no longer changing from `Sublime Text 4` to `Sublime Text_is1` so this will have to be fixed.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ﻿$ErrorActionPreference = 'Stop'
 
+$softwareName = 'Sublime Text*'
```
Comment (Jackenmen) [inline automatic/sublimetext4/tools/chocolateyInstall.ps1]: Hmm, I feel like we shouldn't be matching Sublime Text 3 here, considering that it's a separate package on Chocolatey.
It also appears to be possible to run them side-by-side:
https://www.sublimetext.com/docs/side_by_side.html

I'm not sure if we can phrase this with a single `-like` pattern though. Perhaps we should disable auto-uninstall altogether with `.skipAutoUninstall` file and just handle uninstall in `chocolateyUninstall.ps1` by trying 2 different registry key patterns - "Sublime Text 4" and "Sublime Text_is1" - without any wildcards.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 [hunk 3]
```diff
@@ -8,7 +10,23 @@ $packageArgs = @{
...
+    Install-BinFile -Name 'subl' -Path $sublInstallLocation
+  }
+} elseif ($key.Count -eq 0) {
+  Write-Warning "Not installing subl.exe shim, unable to determine install location"
+} elseif ($key.Count -gt 1) {
+  Write-Warning "$($key.Count) matches found!"
+  Write-Warning "Not installing subl.exe shim, too many install locations found"
+  Write-Warning "Please alert package maintainer the following keys were matched:"
+  $key | ForEach-Object {Write-Warning "- $($_.InstallLocation)"}
+}
```
Comment (Jackenmen) [inline automatic/sublimetext4/tools/chocolateyInstall.ps1]: Per other review comments, this code will probably need to be used by both `chocolateyInstall.ps1` and `chocolateyUninstall.ps1` so it would be great to split it into a separate function in a `helpers.ps1` file.

**Review by Jackenmen — CHANGES_REQUESTED** [review]
I like the idea of getting uninstall registry key to figure out the install location. I left a couple of review comments, note that if you don't want to take on all of these new issues yourself, I can work on them myself (since I have to fix its bugginess eventually anyway) and let you rebase on top of it later, just let me know :)

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 (hunk 1 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyUninstall.ps1:5]: I spent a lot of time on this and wasn't able to determine how to turn off Chocolatey's auto-uninstaller.  It's not documented on their site.  But of course, [a post on reddit from one of their devs](https://stackoverflow.com/questions/45871385/is-there-a-way-to-disable-autouninstaller-from-uninstaller-file) explains how, and you mentioning the `.skipAutoUninstall` file.  :)

That said: I don't know at what point a `.skipAutoUninstall` file has to be put in place.  Is it part of the git repo under tools or mentioned in the nuspec?  Is it done in chocolateyUninstall.ps1?  Is it done in chocolatelyBeforeModify.ps1?

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 (hunk 2 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: I wish I knew at what point/what ST4 version Sublime HQ Pty Ltd did this (re: changing the string from `Sublime Text 4` to `Sublime Text`), since it may very well be that nobody has the non-4 string in use any more, at least via Chocolatey.

The `_is1` suffix appears statically defined, according to [this reddit post](https://stackoverflow.com/questions/35156309/remove-is1-suffix-from-inno-setup-appid), which references the Inno Setup installer code.

`-like` has basic wildcard and range support, but no exclusion support (unlike `-match` which is regex-based).  Docs: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_wildcards?view=powershell-7.3

In short: probably need to check both registry keys without any wildcards, as as you said.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 (hunk 3 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: The behaviours of the `ForEach-Object` code block differs, as do the `Write-Warning` strings, depending on install vs. uninstall context.  I don't know how to code a single helper function to handle both scenarios.  A big problem is I'm not a PowerShell expert (I live more in the C and Python world, and in UNIX space, not so much Windows).  :)

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 (hunk 1 above)
Comment (Jackenmen) [reply automatic/sublimetext4/tools/chocolateyUninstall.ps1:5]: I would just put it as part of the repo.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 (hunk 2 above)
Comment (Jackenmen) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: I'm pretty sure I tested the uninstall script when I made the package so the old name should be present in the very first version pushed to chocolatey, at the very least. If you want to try it out then that's the version I would suggest to check.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 (hunk 3 above)
Comment (Jackenmen) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: That's fine, if you do it separately I'll play with it myself and either figure it out or give up and leave it separate :) I'm not powershell expert either so we'll see.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 (hunk 2 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: It just dawned on me: do you happen to know what `softwareName` even does on an **install**?  On an uninstall I understand the relevance, but not on an install.  The Chocolatey docs<sup>[[1]](https://docs.chocolatey.org/en-us/create/functions/install-chocolateypackage)</sup><sup>[[2]](https://docs.chocolatey.org/en-us/create/functions/install-chocolateyinstallpackage)</sup> don't explain it, and [the code](https://github.com/chocolatey/choco/blob/master/src/chocolatey.resources/helpers/functions/Install-ChocolateyInstallPackage.ps1) doesn't even reference it.

Comment (Jackenmen) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: I assumed that it's used by auto uninstaller but it either isn't the case or is only the case in some edge cases since auto uninstaller uninstalls Sublime Text fine despite incorrect pattern being used.

Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: I found https://github.com/pauby/ChocoPackages/blob/master/automatic/webview2-runtime/tools/chocolateyinstall.ps1 which has the comment:

> #part or all of the Display Name as you see it in Programs and Features. It should be enough to be unique

So I think for the **install**, just saying `Sublime Text` is enough, since at least in Build 4143, that is the literal string that shows up for me in Programs and Features/list of programs to uninstall.  There is no version or build string or architecture (x86/x64) at the end.

Ironically, the uninstaller script for the same package -- https://github.com/pauby/ChocoPackages/blob/master/automatic/webview2-runtime/tools/chocolateyuninstall.ps1 -- does not use a wildcard at all.  This is just another example of the incongruities I find throughout Chocolatey stuff, even when the package maintainer is a Chocolatey engineering manager.  :)

In short: I'm not sure we need to do a double-registry-check at all.  Say people are using sublimetext4 Choco package 4.0.0.1234 (just making up a number here).  If they do a `choco upgrade sublimetext4`, this will get them 4.0.0.414300.  The chocoUninstall.ps1 that runs, prior to the upgrade, would be the one for 4.0.0.1234, which **would not** contain the changes we're talking about here.  It would be the one from whenever you packed + pushed 4.0.0.1234.

Comment (Jackenmen) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: Oh yeah, you're right that we don't need to have a pattern for "Sublime Text 4" since at the point we get the key in either of scripts, we should expect to have new enough version of Sublime Text 4 for it to use the new "Sublime Text_is1" name since that's what the package installed.
I don't want the pattern to match Sublime Text 3 though so we definitely will want to include `_is1` in the pattern that we pass to `Get-UninstallRegistryKey`. As for softwareName parameter to Install-ChocolateyPackage, just remove it entirely, it turns out it's no longer used by the function which you can actually see in its signature in the documentation, it's just that the examples there are outdated.

Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: Ah yes! We definitely want `_is1` here, I forgot to mention that.

As for `softwareName` in the installer: yep, figured as much (looks like the parameter ends up in the "ignored params" section).  Great, I'll clean that up too.

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 [hunk 4]
```diff
@@ -1,26 +1,29 @@
...
-      file          = "$($_.UninstallString.Trim('"'))"
+      packageName    = $packageName
+      fileType       = 'exe'
+      silentArgs     = '/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /TASKS="contextentry"'
+      validExitCodes = @(0)
+      file           = "$($_.UninstallString.Trim('"'))"
+      softwareName   = $softwareName
     }
- 
+    Uninstall-BinFile -Name 'subl'
```
Comment (Jackenmen) [inline automatic/sublimetext4/tools/chocolateyUninstall.ps1]: I think we should always remove the shim, regardless of whether we manage to find a single registry key or not.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 [hunk 5]
```diff
@@ -1,14 +1,33 @@
 ﻿$ErrorActionPreference = 'Stop'
 
+$packageName = $env:ChocolateyPackageName
+$softwareName = 'Sublime Text'
```
Comment (Jackenmen) [inline automatic/sublimetext4/tools/chocolateyInstall.ps1:4]:
```suggestion
$softwareName = 'Sublime Text_is1'
```

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 [hunk 6]
```diff
@@ -1,26 +1,29 @@
-$ErrorActionPreference = 'Stop'
+﻿$ErrorActionPreference = 'Stop'
 
-$packageName = "sublimetext4"
+$packageName = $env:ChocolateyPackageName
+$softwareName = 'Sublime Text'
```
Comment (Jackenmen) [inline automatic/sublimetext4/tools/chocolateyUninstall.ps1:6]:
```suggestion
$softwareName = 'Sublime Text_is1'
```

**Review by Jackenmen — CHANGES_REQUESTED** [review]

**Comment by koitsu** [comment]
OK, 8d0ef9cc9028704a261c285ebc21d30e21c4643f should address most of this.  It brings back the previous uninstaller methodology, and leverages `.skipAutoUninstaller` like discussed.

The first thing you're going to say is "Wait! $softwareName is `Sublime Text` not `Sublime Text_is1` like we discussed!" Yep, you're right. Turns out what Chocolatey's `Get-UninstallRegistryKey` does is recursively look through all the uninstaller registry entries' `DisplayName` fields, not the uninstaller registry key path/folder itself (the `Sublime Text_is1` part of `HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Sublime Text_is1`: https://github.com/chocolatey/choco/blob/master/src/chocolatey.resources/helpers/functions/Get-UninstallRegistryKey.ps1#L137

Now we're back to the conversation about PowerShell `-like`.  `-like` only matches substrings if you use wildcards, otherwise it's equivalent to an `==`.  In other words: `Sublime Text` shouldn't match `Sublime Text 3`, but `Sublime Text*` certainly would.

I **have not** tested installing `SublimeText3` package alongside `sublimetext4`.  I will try to test that tomorrow.

**Comment by Jackenmen** [comment]
> The first thing you're going to say is "Wait! $softwareName is `Sublime Text` not `Sublime Text_is1` like we discussed!" Yep, you're right. Turns out what Chocolatey's `Get-UninstallRegistryKey` does is recursively look through all the uninstaller registry entries' `DisplayName` fields, not the uninstaller registry key path/folder itself

Ah, that makes sense

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 (hunk 4 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyUninstall.ps1]: 64f928cf384e41dc71ebdedfa9162fa1333a62cb should take care of that.

File: automatic/sublimetext4/tools/chocolateyInstall.ps1 (hunk 2 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyInstall.ps1]: Discussed further below, resolving this convo.

File: automatic/sublimetext4/tools/chocolateyUninstall.ps1 (hunk 1 above)
Comment (koitsu) [reply automatic/sublimetext4/tools/chocolateyUninstall.ps1:5]: Fixed in as part of 8d0ef9cc9028704a261c285ebc21d30e21c4643f

**Comment by koitsu** [comment]
Quickly checked out SublimeText3 (package maintained by paulby) and registry bits after install:

- Reg key: `HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Sublime Text 3_is1`
- DisplayName: `Sublime Text 3`
- ST3 **does** include `subl.exe` but the `SublimeText3` package *does not* install a shim for it, so there's no conflict if both packages are installed

Other stuff I just tested ("ST4" below means the latest sublimetext4 package being tested as part of this PR):

- Installed ST3 then ST4: no problems
- Uninstalled ST4: ST3 continued to function fine
- Reinstalled ST4
- Uninstalled ST3: ST4 continued to function fine.  However, what actually happened with ST3 was amusing:
  - Chocolatey emitted warnings + dumping the uninstall registry key entries for both ST3 **and ST4**
    - This happened because the SublimeText3 package erroneously uses `sublime text*` as its `softwareName`: https://github.com/pauby/ChocoPackages/blob/master/automatic/sublimetext3/tools/chocolateyUninstall.ps1
    - I should probably submit a PR to paulby to get that changed to `Sublime Text 3`, no wildcards
    - I don't think this problem has anything to do with the changes in ST4 :)
  - Chocolately emitted warning "To prevent accidental data loss, no programs will be uninstalled" message
  - Chocolatey's auto-uninstaller ran anyway and removed ST3 from the system (and Chocolatey) but **did not** touch ST4

So I think overall we're good here.

**Review by Jackenmen — APPROVED** [review]
Looks great, sorry for the delay. This will become available once a new version of Sublime Text 4 is released.
